ray_on_golem/server/cluster/nodes.py

```diff
@@ -0,0 +1,381 @@
+import asyncio
+import logging
+from datetime import timedelta
+from pathlib import Path
+from typing import TYPE_CHECKING, Collection, List, Optional, Tuple
+
+from golem.exceptions import GolemException
+from golem.managers.base import ManagerException, WorkContext
+from golem.resources import Activity
+from golem.utils.asyncio import create_task_with_logging, ensure_cancelled
+from golem.utils.logging import get_trace_id_name
+
+from ray_on_golem.exceptions import RayOnGolemError
+from ray_on_golem.server.cluster.sidecars import (
+    ActivityStateMonitorClusterNodeSidecar,
+    ClusterNodeSidecar,
+    HeadNodeToWebserverTunnelClusterNodeSidecar,
+    SshStateMonitorClusterNodeSidecar,
+)
+from ray_on_golem.server.mixins import WarningMessagesMixin
+from ray_on_golem.server.models import NodeConfigData, NodeData, NodeState
+from ray_on_golem.server.services.golem.manager_stack import ManagerStack
+from ray_on_golem.server.services.new_golem import GolemService
+from ray_on_golem.server.settings import (
+    CLUSTER_MONITOR_CHECK_INTERVAL,
+    CLUSTER_MONITOR_RETRY_COUNT,
+    CLUSTER_MONITOR_RETRY_INTERVAL,
+)
+from ray_on_golem.utils import get_ssh_command, get_ssh_command_args, run_subprocess_output
+
+if TYPE_CHECKING:
+    from ray_on_golem.server.cluster import Cluster
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterNode(WarningMessagesMixin, NodeData):
+    node_config: NodeConfigData
+    ssh_private_key_path: Path
+    ssh_public_key_path: Path
+    ssh_user: str
+
+    _cluster: "Cluster"
+    _golem_service: GolemService
+    _manager_stack: ManagerStack
+
+    _sidecars: Collection[ClusterNodeSidecar]
+    _ssh_public_key_data: str
+    _activity: Optional[Activity] = None
+    _start_task: Optional[asyncio.Task] = None
+    _warning_messages: List[str] = None
+
+    def __init__(
+        self, cluster: "Cluster", golem_service: GolemService, manager_stack: ManagerStack, **kwargs
+    ) -> None:
+        super().__init__(**kwargs)
+
+        self._cluster = cluster
+        self._golem_service = golem_service
+        self._manager_stack = manager_stack
+
+        self._sidecars = self._prepare_sidecars()
+
+        with self.ssh_public_key_path.open() as f:
+            self._ssh_public_key_data = f.readline().strip()
+
+    def __str__(self) -> str:
+        return self.node_id
```

Comment:
```suggestion
        return f"{self.__class__.__name__}: {self.node_id}"
```

Reply:
The `__str__` is aimed to be as simple as possible, and it is not used as `__repr__`.